shared/route-tree/index.js

```diff
@@ -0,0 +1,229 @@
+// don't exist along the way.
+function _routeSet (routeDef: RouteDefNode, routeState: ?RouteStateNode, pathSpec: PathSetSpec): RouteStateNode {
+  const pathHead = pathSpec && pathSpec.first()
+
+  let newRouteState = routeState || new RouteStateNode({selected: routeDef.defaultSelected})
+  if (pathHead && pathHead.type === 'navigate') {
+    newRouteState = newRouteState.set('selected', pathHead.next || routeDef.defaultSelected)
+    if (pathHead.next === null) {
+      newRouteState = newRouteState.delete('children')
+    }
+  }
+
+  const childName = pathHead && pathHead.type === 'next' ? pathHead.next : newRouteState.selected
+  if (childName !== null) {
+    const childDef = routeDef.getChild(childName)
+    if (!childDef) {
+      throw new InvalidRouteError(`Invalid route child: ${childName}`)
+    }
+
+    newRouteState = newRouteState.updateChild(childName, childState => {
+      let newChild = _routeSet(childDef, childState, pathSpec.skip(1))
+      if (pathHead && pathHead.hasOwnProperty('props')) {
+        newChild = newChild.set('props', I.fromJS(pathHead.props))
+      }
+      return newChild
+    })
+  }
+
+  return newRouteState
+}
+
+export function routeSetProps (routeDef: RouteDefNode, routeState: ?RouteStateNode, pathProps: PathParam, parentPath: ?Path): RouteStateNode {
+  const pathSeq = I.Seq(pathProps).map(item => {
+    if (typeof item === 'string') {
+      return {type: 'navigate', next: item}
+    } else {
+      const {selected, ...props} = item
+      return {type: 'navigate', next: selected, props}
+    }
+  })
+  const parentPathSeq = I.Seq(parentPath || []).map(item => {
+    return {type: 'next', next: item}
+  })
+  return _routeSet(routeDef, routeState, parentPathSeq.concat(pathSeq))
+}
+
+export function routeNavigate (routeDef: RouteDefNode, routeState: ?RouteStateNode, pathProps: PathParam, parentPath: ?Path): RouteStateNode {
+  return routeSetProps(routeDef, routeState, I.List(pathProps).push({selected: null}), parentPath)
+}
+
+export function routeSetState (routeDef: RouteDefNode, routeState: RouteStateNode, path: Path, partialState: {}): RouteStateNode {
+  const pathSeq = I.Seq(path)
+  if (!pathSeq.size) {
+    return routeState.update('state', state => state.merge(partialState))
+  }
+  return routeState.updateChild(pathSeq.first(),
+    childState => routeSetState(routeDef, childState, pathSeq.skip(1), partialState)
+  )
+}
+
+export function routeClear (routeState: ?RouteStateNode, path: Path): ?RouteStateNode {
+  if (!routeState) {
+    return null
+  }
+  const pathSeq = I.Seq(path)
+  if (!pathSeq.size) {
+    return null
+  }
+  return routeState.updateChild(pathSeq.first(),
+    childState => routeClear(childState, pathSeq.skip(1))
+  )
+}
+
+// Traverse a routeState, making sure it matches the routeDef and ends on a leaf component.
+export function checkRouteState (routeDef: RouteDefNode, routeState: ?RouteStateNode): ?string {
+  if (!routeDef) {
+    return
```

Comment:
I think we need this for the sanity check to noop before the initial `setRouteDefs` action is emitted. Not having routes defined is not an error state. It just means we're uninitialized.

However, now I'm looking at this code again, and checking the route state after every action is a bad idea. We should filter the sanity check in https://github.com/keybase/client/pull/4857/files/d2f8e217835f9eac451ddc2644ca1f336086c6c0#diff-7bcbdd8845e7fd8f60a9698dd85b4cbcR102 to only run after routeTree actions, and can probably remove this case after doing so.